gno.land/cmd/genesis/balances_add.go

```diff
@@ -0,0 +1,346 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/gnolang/gno/gno.land/pkg/gnoland"
+	"github.com/gnolang/gno/tm2/pkg/amino"
+	"github.com/gnolang/gno/tm2/pkg/bft/types"
+	"github.com/gnolang/gno/tm2/pkg/commands"
+	"github.com/gnolang/gno/tm2/pkg/sdk/bank"
+	"github.com/gnolang/gno/tm2/pkg/std"
+
+	_ "github.com/gnolang/gno/gno.land/pkg/sdk/vm"
+)
+
+var (
+	balanceRegex = regexp.MustCompile(`^([A-Za-z0-9]+)=([0-9]+)ugnot$`)
+	amountRegex  = regexp.MustCompile(`^(\d+)ugnot$`)
+)
+
+var (
+	errNoBalanceSource        = errors.New("at least one balance source must be set")
+	errMultipleBalanceSources = errors.New("only one mode can be set at a time")
+	errBalanceParsingAborted  = errors.New("balance parsing aborted")
+)
+
+type balancesAddCfg struct {
+	rootCfg *balancesCfg
+
+	balanceSheet  string
+	singleEntries commands.StringArr
+	parseExport   string
+}
+
+// newBalancesAddCmd creates the genesis balances add subcommand
+func newBalancesAddCmd(rootCfg *balancesCfg, io *commands.IO) *commands.Command {
+	cfg := &balancesAddCfg{
+		rootCfg: rootCfg,
+	}
+
+	return commands.NewCommand(
+		commands.Metadata{
+			Name:       "add",
+			ShortUsage: "balances add [flags]",
+			LongHelp:   "Adds a new validator to the genesis.json",
+		},
+		cfg,
+		func(ctx context.Context, _ []string) error {
+			return execBalancesAdd(ctx, cfg, io)
+		},
+	)
+}
+
+func (c *balancesAddCfg) RegisterFlags(fs *flag.FlagSet) {
+	fs.StringVar(
+		&c.balanceSheet,
+		"balance-sheet",
+		"",
+		"the path to the balance file containing addresses in the format <address>=<amount>ugnot",
+	)
+
+	fs.Var(
+		&c.singleEntries,
+		"single",
+		"the direct balance addition in the format <address>=<amount>ugnot",
+	)
+
+	fs.StringVar(
+		&c.parseExport,
+		"parse-export",
+		"",
+		"the path to the transaction export containing a list of transactions (JSONL)",
+	)
+}
+
+func execBalancesAdd(ctx context.Context, cfg *balancesAddCfg, io *commands.IO) error {
+	// Load the genesis
+	genesis, loadErr := types.GenesisDocFromFile(cfg.rootCfg.genesisPath)
+	if loadErr != nil {
+		return fmt.Errorf("unable to load genesis, %w", loadErr)
+	}
+
+	// Validate the source is set correctly
+	var (
+		singleEntriesSet = len(cfg.singleEntries) != 0
+		balanceSheetSet  = cfg.balanceSheet != ""
+		txFileSet        = cfg.parseExport != ""
+	)
+
+	if err := validateSetModes(
+		[]bool{
+			singleEntriesSet,
+			balanceSheetSet,
+			txFileSet,
+		},
+	); err != nil {
+		return fmt.Errorf("invalid modes set, %w", err)
+	}
```

Comment:
The idea is that we want to have only 1 enabled at any time, since the user needs to specify a "source" of balances.

I think it's probably more clear to error out if the user sets 2 -- how do we choose which source to parse?

Are you suggesting we parse every enabled source? It's actually totally doable, would simplify this a lot

Reply:
Exactly that, parse every enabled source.

Reply:
I added support for this here:

[cea7a2a](https://github.com/gnolang/gno/pull/1276/commits/cea7a2a6fd1538f368b0727c0587d42ecce22d37)